research/sentiment_analysis/sentiment_main.py

```diff
@@ -0,0 +1,188 @@
+
+  estimator = tf.keras.estimator.model_to_estimator(
+      keras_model=keras_model, model_dir=model_dir, config=run_config)
+
+  return estimator
+
+
+def per_device_batch_size(batch_size, num_gpus):
+  """For multi-gpu, batch-size must be a multiple of the number of GPUs.
+  Note that this should eventually be handled by DistributionStrategies
+  directly. Multi-GPU support is currently experimental, however,
+  so doing the work here until that feature is in place.
+  Args:
+    batch_size: Global batch size to be divided among devices. This should be
+      equal to num_gpus times the single-GPU batch_size for multi-gpu training.
+    num_gpus: How many GPUs are used with DistributionStrategies.
+  Returns:
+    Batch size per device.
+  Raises:
+    ValueError: if batch_size is not divisible by number of devices
+  """
+  if num_gpus <= 1:
+    return batch_size
+
+  remainder = batch_size % num_gpus
+  if remainder:
+    err = ("When running with multiple GPUs, batch size "
+           "must be a multiple of the number of available GPUs. Found {} "
+           "GPUs with a batch size of {}; try --batch_size={} instead."
+          ).format(num_gpus, batch_size, batch_size - remainder)
+    raise ValueError(err)
+  return int(batch_size / num_gpus)
+
+def run_model(_):
+  """Run training and eval loop."""
+
+  tf.logging.info("Creating Estimator from Keras model...")
+  num_class = imdb.NUM_CLASS
+
+  keras_model = sentiment_model.CNN(
+      FLAGS.embedding_dim, FLAGS.vocabulary_size, FLAGS.sentence_length,
+      FLAGS.cnn_filters, num_class, FLAGS.dropout_rate, 0.0)
+  num_gpus = flags_core.get_num_gpus(FLAGS)
+  estimator = convert_keras_to_estimator(keras_model, num_gpus, FLAGS.model_dir)
+
+  # Create hooks that log information about the training and metric values
+  train_hooks = hooks_helper.get_train_hooks(
+      FLAGS.hooks,
+      batch_size=FLAGS.batch_size  # for ExamplesPerSecondHook
+  )
+  run_params = {
+      "batch_size": FLAGS.batch_size,
+      "train_epochs": FLAGS.train_epochs,
+  }
+  benchmark_logger = logger.get_benchmark_logger()
+  benchmark_logger.log_run_info(
+      model_name="sentiment_analysis",
+      dataset_name=FLAGS.dataset,
+      run_params=run_params,
+      test_id=FLAGS.benchmark_test_id)
+
+  # Training and evaluation cycle
+  def train_input_fn():
+    return dataset.input_fn(
+        FLAGS.dataset, True, per_device_batch_size(FLAGS.batch_size, num_gpus),
+        FLAGS.vocabulary_size, FLAGS.sentence_length, repeat=FLAGS.epochs_between_evals)
+
+  total_training_cycle = FLAGS.train_epochs // FLAGS.epochs_between_evals
+
+  for cycle_index in range(total_training_cycle):
+    tf.logging.info("Starting a training cycle: {}/{}".format(
+        cycle_index + 1, total_training_cycle))
+
+    # Train the model
+    estimator.train(input_fn=train_input_fn, hooks=train_hooks)
+
+    # Evaluate the model
+    eval_results = evaluate_model(estimator, num_gpus)
```

Comment:
Why the discrepancy in style where eval gets a helper function but train doesn't?

Reply:
The evaluation process can be moved here, without explicitly creating a function. In recommendation model (if you take it as reference), because the evaluation process is a little bit complex, it's abstracted as a function. Here the evaluate is simple, and can be directly placed here.